os/arch/arm/src/imxrt/imxrt_i2s.c

```diff
@@ -0,0 +1,2707 @@
+#define       i2s_init_buffer(b, s)   memset(b, 0x55, s);
+#define       i2s_dump_buffer(m, b, s) lib_dumpbuffer(m, b, s)
+#else
+#define       i2s_init_buffer(b, s)
+#define       i2s_dump_buffer(m, b, s)
+#endif
+
+/* Semaphore helpers */
+static void i2s_exclsem_take(struct imxrt_i2s_s *priv);
+#define i2s_exclsem_give(priv) sem_post(&priv->exclsem)
+
+/* I2S methods (and close friends) */
+
+static uint32_t i2s_rxdatawidth(struct i2s_dev_s *dev, int bits);
+
+static uint32_t i2s_samplerate(struct i2s_dev_s *dev, uint32_t rate);
+
+static uint32_t i2s_txdatawidth(struct i2s_dev_s *dev, int bits);
+
+static int i2s_stop_transfer(struct i2s_dev_s *dev, i2s_ch_dir_t dir);
+
+static int i2s_stop(struct i2s_dev_s *dev, i2s_ch_dir_t dir);
+static int i2s_pause(struct i2s_dev_s *dev, i2s_ch_dir_t dir);
+static int i2s_resume(struct i2s_dev_s *dev, i2s_ch_dir_t dir);
+
+static int i2s_receive(struct i2s_dev_s *dev, struct ap_buffer_s *apb, i2s_callback_t callback, void *arg, uint32_t timeout);
+static int i2s_send(struct i2s_dev_s *dev, struct ap_buffer_s *apb, i2s_callback_t callback, void *arg, uint32_t timeout);
+
+static int i2s_err_cb_register(struct i2s_dev_s *dev, i2s_err_cb_t cb, void *arg);
+
+int i2s_set_clk(struct imxrt_i2s_s *priv, uint32_t rate, i2s_bits_per_sample_t bits, i2s_channel_t ch);
+
+/****************************************************************************
+* External functions
+****************************************************************************/
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+/* Base pointer array */
+static I2S_Type *const s_saiBases[] = I2S_BASE_PTRS;
+/*!@brief SAI handle pointer */
+sai_handle_t *s_saiHandle[ARRAY_SIZE(s_saiBases)][2];
+
+#if !(defined(FSL_SDK_DISABLE_DRIVER_CLOCK_CONTROL) && FSL_SDK_DISABLE_DRIVER_CLOCK_CONTROL)
+/* Clock name array */
+static const clock_ip_name_t s_saiClock[] = SAI_CLOCKS;
+#endif /* FSL_SDK_DISABLE_DRIVER_CLOCK_CONTROL */
+
+static const i2s_config_t i2s0_default_config = {
+#if defined(CONFIG_IMXRT_I2S0_MODE_MASTER) && (0 < CONFIG_IMXRT_I2S0_MODE_MASTER)
+	.mode = I2S_MODE_MASTER,
+#else
+	.mode = I2S_MODE_SLAVE,
+#endif
+
+	.sample_rate = CONFIG_IMXRT_I2S_SAMPLERATE,
+	.bits_per_sample = CONFIG_IMXRT_I2S_DATALEN,
+	.channel_format = CONFIG_IMXRT_I2S_CHANNEL_FMT,
+	.communication_format = CONFIG_IMXRT_I2S_COMM_FMT,
+	.int_alloc_level = CONFIG_IMXRT_I2S_INTR_LEVEL,
+};
+
+static const i2s_config_t i2s1_default_config = {
+#if defined(CONFIG_IMXRT_I2S1_MODE_MASTER) && (0 < CONFIG_IMXRT_I2S1_MODE_MASTER)
+	.mode = I2S_MODE_MASTER,
+#else
+	.mode = I2S_MODE_SLAVE,
+#endif
+
+	.sample_rate = CONFIG_IMXRT_I2S_SAMPLERATE,
+	.bits_per_sample = CONFIG_IMXRT_I2S_DATALEN,
+	.channel_format = CONFIG_IMXRT_I2S_CHANNEL_FMT,
+	.communication_format = CONFIG_IMXRT_I2S_COMM_FMT,
+	.int_alloc_level = CONFIG_IMXRT_I2S_INTR_LEVEL,
```

Comment:
Is it the i2s interrupt or the dma interrupt associated with the i2s port?

Reply:
It is I2S interrupt.
Actually, we plan to use imxrt_i2s_edma.c file 
That is why eDMA function is implemented for I2S feature